jerry-debugger/src/lib/protocol-handler.ts

```diff
@@ -0,0 +1,251 @@
+// limitations under the License.
+
+import * as SP from './jrs-protocol-constants';
+import { ParsedFunction } from './breakpoint';
+import { ByteConfig, cesu8ToString, concatUint8Arrays, decodeMessage } from './utils';
+
+// expected JerryScript debugger protocol version
+const JERRY_DEBUGGER_VERSION = 1;
+
+export interface ParserStackFrame {
+  isFunc: boolean;
+  line: number;
+  column: number;
+  name: string;
+  source: string;
+  sourceName?: string;
+  lines: Array<number>;
+  offsets: Array<number>;
+  byteCodeCP?: number;
+  firstBreakpointLine?: number;
+  firstBreakpointOffset?: number;
+}
+
+export interface JerryDebugProtocolDelegate {
+  onError?(message: string): void;
+  onScriptParsed?(message: JerryMessageScriptParsed): void;
+}
+
+export interface JerryMessageScriptParsed {
+  scriptId: number;
+  name: string;
+  endLine: number;
+  endColumn: number;
+  hasSourceURL: boolean;
+}
+
+interface ProtocolFunctionMap {
+  [type: number]: (data: Uint8Array) => void;
+}
+
+interface FunctionMap {
+  [cp: string]: ParsedFunction;
+}
+
+// abstracts away the details of the protocol
+export class JerryDebugProtocolHandler {
+  public delegate: JerryDebugProtocolDelegate;
+
+  // debugger configuration
+  private maxMessageSize: number = 0;
+  private byteConfig: ByteConfig;
+  private version: number = 0;
+  private functionMap: ProtocolFunctionMap;
+
+  private stack: Array<ParserStackFrame>;
+  private source: string = '';
+  private sourceData?: Uint8Array;
+  private sourceName?: string;
+  private sourceNameData?: Uint8Array;
+  private functionName?: Uint8Array;
+  private functions: FunctionMap = {};
+  private lastScriptID: number = 0;
+
+  constructor(delegate: JerryDebugProtocolDelegate) {
+    this.delegate = delegate;
+
+    this.byteConfig = {
+      cpointerSize: 0,
+      littleEndian: true,
+    };
+
+    this.functionMap = {
+      [SP.JERRY_DEBUGGER_CONFIGURATION]: this.onConfiguration,
+      [SP.JERRY_DEBUGGER_BYTE_CODE_CP]: this.onByteCodeCP,
+      [SP.JERRY_DEBUGGER_PARSE_FUNCTION]: this.onParseFunction,
+      [SP.JERRY_DEBUGGER_BREAKPOINT_LIST]: this.onBreakpointList,
+      [SP.JERRY_DEBUGGER_BREAKPOINT_OFFSET_LIST]: this.onBreakpointList,
+      [SP.JERRY_DEBUGGER_SOURCE_CODE]: this.onSourceCode,
+      [SP.JERRY_DEBUGGER_SOURCE_CODE_END]: this.onSourceCode,
+      [SP.JERRY_DEBUGGER_SOURCE_CODE_NAME]: this.onSourceCodeName,
+      [SP.JERRY_DEBUGGER_SOURCE_CODE_NAME_END]: this.onSourceCodeName,
+    };
+
+    this.stack = [{
+      isFunc: false,
+      line: 1,
+      column: 1,
+      name: '',
+      source: '',
+      lines: [],
+      offsets: [],
+    }];
+  }
+
+  // for the temporary mollification of lint
+  unused() {
+    this.maxMessageSize;
+  }
+
+  decodeMessage(format: string, message: Uint8Array, offset: number) {
+    return decodeMessage(this.byteConfig, format, message, offset);
+  }
+
+  onConfiguration(data: Uint8Array) {
```

Comment:
> I realize for one thing my onFunctions are public, not sure if that's right - but it is convenient in that I can write unit tests for them. 

Re. public: we've been using the idea of "cracker" subclasses in our code. The idea is simple: use `protected` and then declare a "cracker" subclass in the test that changes the visibility from `protected` to `public`.

```
// impl.ts
class JerryDebugProtocolHandler {
  protected onConfiguration(data: Uint8Array) { ... }
  protected onByteCodeCP(data: Uint8Array) { ... }
  // ...etc
}

// ....test.ts:
class JerryDebugProtocolHandlerCracker extends JerryDebugProtocolHandler {
  public onConfiguration(data: Uint8Array);
  public onByteCodeCP(data: Uint8Array);
  // ...etc
}

// using it -- just cast to the cracker type:
const handler = new JerryDebugProtocolHandler(delegate) as JerryDebugProtocolHandlerCracker;
handler.onConfiguration; // this is legal now

```

Reply:
OK, I'll try to keep that in mind when I notice  this again but I'm not going to scour the code for it right this moment.